### PR TITLE
feat: Implement custom source formatting to improve performance

### PR DESCRIPTION
**Pull request by latonz**
# Implement custom source formatting to improve performance

## Description

Implements custom source formatting during the creation of the syntax nodes.
This PR slightly changes the formatting of a few things to improve performance / readability:
* After a block no empty line is inserted
* Initializer blocks: each assignment is on its own line (even for older Roslyn versions) and has a trailing comma

<details>
  <summary>Performance current main</summary>

|       Method |        Mean |     Error |    StdDev |      Gen0 |     Gen1 |   Allocated |
|------------- |------------:|----------:|----------:|----------:|---------:|------------:|
|      Compile |    582.8 us |  11.66 us |  29.24 us |   48.8281 |  11.7188 |   302.52 KB |
| LargeCompile | 31,990.4 us | 312.57 us | 277.08 us | 2125.0000 | 875.0000 | 13270.94 KB |

</details>

<details>
  <summary>Performance with this PR</summary>

|       Method |        Mean |     Error |    StdDev |      Gen0 |     Gen1 |   Allocated |
|------------- |------------:|----------:|----------:|----------:|---------:|------------:|
|      Compile |    470.7 us |   9.38 us |  13.45 us |   47.8516 |   6.8359 |   298.45 KB |
| LargeCompile | 19,770.2 us | 187.93 us | 175.79 us | 1906.2500 | 531.2500 | 11816.03 KB |

</details>

Improvement of about 38% for our LargeCompile case and small memory improvements 🚀 
The PR is pretty big, although a lot of it are just reformatted test files and namespace moves.

## Checklist

- [x] The existing code style is followed
- [x] The commit message follows our guidelines
- [x] Performed a self-review of my code
- [x] Hard-to-understand areas of my code are commented
- [x] The documentation is updated (as applicable)
- [x] Unit tests are added/updated
- [x] Integration tests are added/updated (as applicable, especially if feature/bug depends on roslyn or framework version in use)

### CODE REVIEW

File: test/Riok.Mapperly.IntegrationTests/Riok.Mapperly.IntegrationTests.csproj [hunk 1]
```diff
@@ -44,4 +44,10 @@
       <IncludeAssets>runtime; build; native; contentfiles; analyzers</IncludeAssets>
     </PackageReference>
   </ItemGroup>
+  <ItemGroup>
```
Comment (TimothyMakkison) [inline test/Riok.Mapperly.IntegrationTests/Riok.Mapperly.IntegrationTests.csproj:47]: Is this needed? iirc I'd accidentally add a similar block when deleting unneeded tests.

File: src/Riok.Mapperly/Emit/SourceEmitter.cs [hunk 2]
```diff
@@ -12,50 +13,76 @@ public static class SourceEmitter
...
+            ctx = ctx.RemoveIndentation();
             containingType = containingType.ContainingType;
         }
 
         return syntax;
     }
 
-    private static MemberDeclarationSyntax WrapInNamespaceIfNeeded(string? namespaceName, MemberDeclarationSyntax classDeclaration)
+    private static MemberDeclarationSyntax WrapInNamespaceIfNeeded(
+        ref SourceEmitterContext ctx,
```
Comment (TimothyMakkison) [inline src/Riok.Mapperly/Emit/SourceEmitter.cs:67]: Unneeded pass by ref?

File: src/Riok.Mapperly/Helpers/AnalyzerConfigOptionsExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,11 @@
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Riok.Mapperly.Helpers;
+
+public static class AnalyzerConfigOptionsExtensions
```
Comment (TimothyMakkison) [inline src/Riok.Mapperly/Helpers/AnalyzerConfigOptionsExtensions.cs:5]: Is this used anywhere?

File: src/Riok.Mapperly/Emit/Syntax/SyntaxIndentationExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    public static SyntaxToken AddTrailingLineFeed(this SyntaxToken token, int indentation)
+    {
+        var trivia = token.TrailingTrivia.AddLineFeedAndIndentation(indentation);
+        return token.WithTrailingTrivia(trivia);
+    }
+
+    public static SyntaxTriviaList AddLineFeedAndIndentation(this SyntaxTriviaList trivia, int indentation)
+    {
+        trivia = trivia.Insert(0, ElasticCarriageReturnLineFeed);
+        for (var i = 0; i < indentation; i++)
```
Comment (TimothyMakkison) [inline src/Riok.Mapperly/Emit/Syntax/SyntaxIndentationExtensions.cs:87]: could `trivia.InsertRange` be user here? Roslyn is usually Immutable so this is likely allocating a lot of times. Plus inserting to the front of a list causes everything to copied as it is shifted.

**Review by TimothyMakkison — CHANGES_REQUESTED** [review]
Really impressive stuff, a 38% improvement is crazy good, it's probably on par with a string builder version. How long did this take 😨 

`SourceEmitter.Build` now takes 15% of the time. Iirc used to take 50%-60% of the runtime.
image


I haven't read the whole thing but I've added a couple of comments

File: src/Riok.Mapperly/Helpers/StackExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,12 @@
+namespace Riok.Mapperly.Helpers;
+
+public static class StackExtensions
+{
+    public static IEnumerable<T> PopAll<T>(this Stack<T> stack)
```
[the remark this answers is not in the document]
Comment (latonz) [reply src/Riok.Mapperly/Helpers/StackExtensions.cs:5]: Removed

File: src/Riok.Mapperly/Emit/Syntax/SyntaxIndentationExtensions.cs (hunk 4 above)
Comment (latonz) [reply src/Riok.Mapperly/Emit/Syntax/SyntaxIndentationExtensions.cs]: Added a cached trivia array / insert approach, slightly improved the benchmark.

**Comment by latonz** [comment]
@TimothyMakkison thanks for you feedback 😊 
The changes on the initializer blocks are intentional and not adding an empty line after code blocks is an intentional performance wise decision. We could add it to *every* code block pretty easily. But the NormalizeWhitespace implementation only adds it if the next line is not a closing block `}`. This makes it hard as we cannot add the empty line based on the information of the node itself but would need to analyse the nodes afterwards. And adding it after each block looks strange...

**Review by ni507 — APPROVED** [review]

**Pull request merged**